lib/web/fetch/formdata-parser.js

```diff
@@ -0,0 +1,446 @@
+
+      // 5.11.1. Let value be the UTF-8 decoding without BOM of body.
+      value = utf8DecodeBytes(Buffer.from(body))
+    }
+
+    // 5.12. Assert: name is a scalar value string and value is either a scalar value string or a File object.
+    assert(name === webidl.converters.USVString(name))
+    assert((typeof value === 'string' && value === webidl.converters.USVString(value)) || isFileLike(value))
+
+    // 5.13. Create an entry with name and value, and append it to entry list.
+    entryList.push(makeEntry(name, value, filename))
+  }
+}
+
+/**
+ * @see https://andreubotella.github.io/multipart-form-data/#parse-multipart-form-data-headers
+ * @param {Buffer} input
+ * @param {{ position: number }} position
+ */
+function parseMultipartFormDataHeaders (input, position) {
+  // 1. Let name, filename and contentType be null.
+  let name = null
+  let filename = null
+  let contentType = null
+  let encoding = null
+
+  // 2. While true:
+  while (true) {
+    // 2.1. If position points to a sequence of bytes starting with 0x0D 0x0A (CR LF):
+    if (input[position.position] === 0x0d && input[position.position + 1] === 0x0a) {
+      // 2.1.1. If name is null, return failure.
+      if (name === null) {
+        return 'failure'
+      }
+
+      // 2.1.2. Return name, filename and contentType.
+      return { name, filename, contentType, encoding }
+    }
+
+    // 2.2. Let header name be the result of collecting a sequence of bytes that are
+    //      not 0x0A (LF), 0x0D (CR) or 0x3A (:), given position.
+    let headerName = collectASequenceOfBytes(
+      (char) => char !== 0x0a && char !== 0x0d && char !== 0x3a,
+      input,
+      position
+    )
+
+    // 2.3. Remove any HTTP tab or space bytes from the start or end of header name.
+    headerName = removeChars(headerName, true, true, (char) => char === 0x9 || char === 0x20)
+
+    // 2.4. If header name does not match the field-name token production, return failure.
+    if (!HTTP_TOKEN_CODEPOINTS.test(headerName.toString())) {
+      return 'failure'
+    }
+
+    // 2.5. If the byte at position is not 0x3A (:), return failure.
+    if (input[position.position] !== 0x3a) {
+      return 'failure'
+    }
+
+    // 2.6. Advance position by 1.
+    position.position++
+
+    // 2.7. Collect a sequence of bytes that are HTTP tab or space bytes given position.
+    //      (Do nothing with those bytes.)
+    collectASequenceOfBytes(
+      (char) => char === 0x20 || char === 0x09,
+      input,
+      position
+    )
+
+    // 2.8. Byte-lowercase header name and switch on the result:
+    switch (new TextDecoder().decode(headerName).toLowerCase()) {
+      case 'content-disposition': {
+        // 1. Set name and filename to null.
+        name = filename = null
+
+        // 2. If position does not point to a sequence of bytes starting with
+        //    `form-data; name="`, return failure.
+        if (input.indexOf(formDataNameBuffer, position.position) !== position.position) {
```

Comment:
worst case scenario it will go through the whole buffer to find the sequence. using a function like the following avoids that scenario.

```js

/**
 * @param {Buffer} buf 
 * @param {number} idx
 * @returns {boolean}
 */
function isFormDataNameBufferCompare(buf, idx) {
  return (
    (idx + 17) <= buf.length &&
    buf[idx] === 0x66 && // f
    buf[idx + 1] === 0x6f && // o
    buf[idx + 2] === 0x72 && // r
    buf[idx + 3] === 0x6d && // m
    buf[idx + 4] === 0x2d && // -
    buf[idx + 5] === 0x64 && // d
    buf[idx + 6] === 0x61 && // a
    buf[idx + 7] === 0x74 && // t
    buf[idx + 8] === 0x61 && // a
    buf[idx + 9] === 0x3b && // ;
    buf[idx + 10] === 0x20 && // <SP>
    buf[idx + 11] === 0x6e && // n
    buf[idx + 12] === 0x61 && // a
    buf[idx + 13] === 0x6d && // m
    buf[idx + 14] === 0x65 && // e
    buf[idx + 15] === 0x3d && // =
    buf[idx + 16] === 0x22 // "
  )
}
```

Reply:
I came up with a better, more reusable solution